Source/Core/DiscIO/FileSystemGCWii.cpp

```diff
@@ -19,83 +21,309 @@
+{
+	return std::unique_ptr<IFileInfo>(new CFileInfoGCWii(*this, IsDirectory() ? GetSize() : m_index + 1));
+}
+
+IFileInfo::const_iterator CFileInfoGCWii::begin() const
+{
+	return const_iterator(std::unique_ptr<IFileInfo>(new CFileInfoGCWii(*this, m_index + 1)));
+}
+
+IFileInfo::const_iterator CFileInfoGCWii::end() const
+{
+	return const_iterator(Next());
+}
+
+// Reads the first u32 of this entry for Get(0), reads the u32 after that for Get(1), and so on.
+// Each entry contains 3 u32s (0xC bytes).
+u32 CFileInfoGCWii::Get(int n) const { return Common::swap32(*((u32*)(m_fst + m_index * 0xC) + n)); }
+u32 CFileInfoGCWii::GetNameOffset() const { return Get(0); }
+u32 CFileInfoGCWii::GetRawOffset() const { return Get(1); }
+u32 CFileInfoGCWii::GetSize() const
+{
+	u32 result = Get(2);
+
+	if (IsDirectory() && result <= m_index)
+	{
+		// For directories, GetSize is supposed to return the index of the next entry.
+		// If a file system is corrupt and instead has an index that isn't after this one,
+		// we act as if the directory is empty to avoid strange behavior.
+		ERROR_LOG(DISCIO, "Invalid folder end in file system");
+		return m_index + 1;
+	}
+
+	return result;
+}
+
+u64 CFileInfoGCWii::GetOffset() const { return (u64)GetRawOffset() << (m_wii ? 2 : 0); }
+bool CFileInfoGCWii::IsDirectory() const { return (GetNameOffset() & 0xFF000000) != 0; }
+u32 CFileInfoGCWii::GetTotalChildren() const { return GetSize() - (m_index + 1); }
+std::string CFileInfoGCWii::GetName() const
+{
+	const u8* name_offset = m_fst + m_total_file_infos * 0xC + (GetNameOffset() & 0xFFFFFF);
+
+	if (name_offset >= m_fst + GetFSTSize())
+	{
+		// To avoid reading beyond the end of the FST, nothing is read if the name offset is too large.
+		ERROR_LOG(DISCIO, "Name offset in file system is too large");
+		return "";
+	}
+
+	// TODO: Should we really always use SHIFT-JIS?
+	// Some names in Pikmin (NTSC-U) don't make sense without it, but is it correct?
+	return SHIFTJISToUTF8((char*)name_offset);
+}
+
+std::string CFileInfoGCWii::GetPath() const
+{
+	// The root entry doesn't have a name
+	if (m_index == 0)
+		return "";
+
+	if (IsDirectory())
+	{
+		// The index of the parent directory is available through GetRawOffset().
+
+		if (GetRawOffset() >= m_index)
+		{
+			// The index of the parent directory is supposed to be smaller than
+			// the current index. If an FST is corrupt and breaks that rule,
+			// there's a risk that parent directory pointers form a loop.
+			// To avoid stack overflows, this method returns.
+			ERROR_LOG(DISCIO, "Invalid parent offset in file system");
+			return "";
+		}
+		return CFileInfoGCWii(*this, GetRawOffset()).GetPath() + GetName() + "/";
+	}
+	else
+	{
+		// The parent directory can be found by searching backwards
+		// for a directory that contains this file.
+
+		CFileInfoGCWii potential_parent(*this, m_index - 1);
+		while (!(potential_parent.IsDirectory() && potential_parent.GetSize() > m_index))
+		{
```

Comment:
Nit: on malformed FST this could crash or loop indefinitely. I would maybe add an additional exit condition & associated error logging.